// node-derpi/dist/api/Comment.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const User_1 = require("./User");
const Consts = require("../util/Consts");
const json2typescript_1 = require("json2typescript");
const DateConverter_1 = require("../util/DateConverter");
const __1 = require("..");
/**
 * Represents a single comment on an image
 *
 * @export
 * @class Comment
 */
let Comment = class Comment {
    /**
     * Represents a single comment on an image
     *
     * @export
     * @class Comment
     */
    constructor() {
        /**
         * The ID of the comment
         *
         * @readonly
         * @type {number}
         * @memberof Comment
         */
        this.id = 0;
        /**
         * The body of the comment
         *
         * @readonly
         * @type {string}
         * @memberof Comment
         */
        this.body = '';
        /**
         * Has the comment been deleted?
         *
         * @readonly
         * @type {boolean}
         * @memberof Comment
         */
        this.deleted = false;
        /**
         * The date the comment was posted on
         *
         * @readonly
         * @type {Date}
         * @memberof Comment
         */
        this.posted = Consts.DEFAULT_DATE;
        /**
         * The name of the user who posted this comment
         *
         * Use this instead of (await author()).name to save an HTTP request and make the Derpi admins happy
         *
         * @readonly
         * @type {string}
         * @memberof Comment
         */
        this.authorName = '';
        /**
         * The internal ID of the image this comment was posted on
         *
         * @readonly
         * @private
         * @type {number}
         * @memberof Comment
         */
        this._image = 0;
    }
    /**
     * Gets the author of this comment
     *
     * @returns {Promise<User>} A Promise wrapping the user that posted this comment
     * @memberof Comment
     */
    author() {
        return __awaiter(this, void 0, void 0, function* () {
            // Part II of the Background Pony saga:
            // comments_home.json does NOT provide me with the user ID of the uploader, just the name
            // Is this a **user** with the name Background Pony #whatever, or is it a real anonymous user??
            // We're just going to assume the latter and hope nobody breaks anything.
            if (this.authorName.match(/Background Pony \#[0-9A-Z]+/)) {
                let user = new User_1.User();
                user.name = this.authorName; // TODO: figure out how to do this better so user.name can be readonly
                return user;
            }
            let user = yield __1.Fetch.fetchUser(this.authorName);
            return user;
        });
    }
    /**
     * Gets the image this comment belongs to
     *
     * @returns {Promise<Image>} A Promise wrapping the image this comment was posted on
     * @memberof Comment
     */
    image() {
        return __awaiter(this, void 0, void 0, function* () {
            return __1.Fetch.fetchImage(this._image);
        });
    }
};
__decorate([
    json2typescript_1.JsonProperty('id', Number),
    __metadata("design:type", Number)
], Comment.prototype, "id", void 0);
__decorate([
    json2typescript_1.JsonProperty('body', String),
    __metadata("design:type", String)
], Comment.prototype, "body", void 0);
__decorate([
    json2typescript_1.JsonProperty('deleted', Boolean),
    __metadata("design:type", Boolean)
], Comment.prototype, "deleted", void 0);
__decorate([
    json2typescript_1.JsonProperty('posted_at', DateConverter_1.DateConverter),
    __metadata("design:type", Date)
], Comment.prototype, "posted", void 0);
__decorate([
    json2typescript_1.JsonProperty('author', String),
    __metadata("design:type", String)
], Comment.prototype, "authorName", void 0);
__decorate([
    json2typescript_1.JsonProperty('image_id', Number),
    __metadata("design:type", Number)
], Comment.prototype, "_image", void 0);
Comment = __decorate([
    json2typescript_1.JsonObject
], Comment);
exports.Comment = Comment;
