// node-derpi/dist/api/TagDetails.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const json2typescript_1 = require("json2typescript");
const URLConverter_1 = require("../util/URLConverter");
const __1 = require("..");
const TagCollection_1 = require("../util/TagCollection");
/**
 * Represents some basic details about a given tag
 *
 * @export
 * @class TagDetails
 */
let TagDetails = class TagDetails {
    /**
     * Represents some basic details about a given tag
     *
     * @export
     * @class TagDetails
     */
    constructor() {
        /**
         * The ID of this tag
         *
         * @readonly
         * @type {number}
         * @memberof TagDetails
         */
        this.id = 0;
        /**
         * The name of this tag
         *
         * @readonly
         * @type {string}
         * @memberof TagDetails
         */
        this.name = '';
        /**
         * The sluggified name of this tag
         *
         * @readonly
         * @see Helpers#sluggify
         * @type {string}
         * @memberof TagDetails
         */
        this.slug = '';
        /**
         * The description of this tag
         *
         * @readonly
         * @type {string}
         * @memberof TagDetails
         */
        this.description = '';
        /**
         * The number of images on this tag
         *
         * @readonly
         * @type {number}
         * @memberof TagDetails
         */
        this.imageCount = 0;
        /**
         * The namespace this tag is in (part before the colon)
         *
         * @readonly
         * @type {string}
         * @memberof TagDetails
         */
        this.namespace = '';
        /**
         * The name of the tag in its namespace (part after the colon)
         *
         * @readonly
         * @type {string}
         * @memberof TagDetails
         */
        this.nameInNamespace = '';
        /**
         * The category of tags this tag is in
         *
         * @readonly
         * @type {string}
         * @memberof TagDetails
         */
        this.category = '';
        /**
         * The image used for spoilering images with this tag
         *
         * @readonly
         * @type {string}
         * @memberof TagDetails
         */
        this.spoilerImage = '';
        /**
         * The internal tag IDs for tags implied by this tag
         *
         * @readonly
         * @private
         * @type {number[]}
         * @memberof TagDetails
         */
        this._impliedTags = [];
        /**
         * The internal tag ID for the tag this tag is aliased to, if any
         *
         * @readonly
         * @private
         * @type {number}
         * @memberof TagDetails
         */
        this._aliasedTo = -1;
    }
    /**
     * Gets a list of tags implied by this tag
     *
     * @returns {Promise<TagCollection>} A Promise wrapping a TagCollection of implied tags
     * @memberof TagDetails
     */
    impliedTags() {
        return __awaiter(this, void 0, void 0, function* () {
            return new TagCollection_1.TagCollection(this._impliedTags);
        });
    }
    /**
     * Gets the tag this tag is an alias of, if it exists
     *
     * @returns {(Promise<Tag | null>)} The aliased tag, or null if none exists
     * @memberof TagDetails
     */
    aliasedTo() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._aliasedTo === -1)
                return null;
            return __1.Fetch.fetchTagByID(this._aliasedTo);
        });
    }
};
__decorate([
    json2typescript_1.JsonProperty('id', Number),
    __metadata("design:type", Number)
], TagDetails.prototype, "id", void 0);
__decorate([
    json2typescript_1.JsonProperty('name', String),
    __metadata("design:type", String)
], TagDetails.prototype, "name", void 0);
__decorate([
    json2typescript_1.JsonProperty('slug', String),
    __metadata("design:type", String)
], TagDetails.prototype, "slug", void 0);
__decorate([
    json2typescript_1.JsonProperty('description', String),
    __metadata("design:type", String)
], TagDetails.prototype, "description", void 0);
__decorate([
    json2typescript_1.JsonProperty('images', Number),
    __metadata("design:type", Number)
], TagDetails.prototype, "imageCount", void 0);
__decorate([
    json2typescript_1.JsonProperty('namespace', String),
    __metadata("design:type", String)
], TagDetails.prototype, "namespace", void 0);
__decorate([
    json2typescript_1.JsonProperty('name_in_namespace', String),
    __metadata("design:type", String)
], TagDetails.prototype, "nameInNamespace", void 0);
__decorate([
    json2typescript_1.JsonProperty('category', String),
    __metadata("design:type", String)
], TagDetails.prototype, "category", void 0);
__decorate([
    json2typescript_1.JsonProperty('spoiler_image_uri', URLConverter_1.URLConverter),
    __metadata("design:type", String)
], TagDetails.prototype, "spoilerImage", void 0);
__decorate([
    json2typescript_1.JsonProperty('implied_tag_ids', [Number]),
    __metadata("design:type", Array)
], TagDetails.prototype, "_impliedTags", void 0);
__decorate([
    json2typescript_1.JsonProperty('aliased_to_id', Number),
    __metadata("design:type", Number)
], TagDetails.prototype, "_aliasedTo", void 0);
TagDetails = __decorate([
    json2typescript_1.JsonObject
], TagDetails);
exports.TagDetails = TagDetails;
