// node-derpi/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Fetch_1 = require("./util/Fetch");
exports.Fetch = Fetch_1.Fetch;
Fetch_1.Fetch.setup();
var Award_1 = require("./api/Award");
exports.Award = Award_1.Award;
var Comment_1 = require("./api/Comment");
exports.Comment = Comment_1.Comment;
var Image_1 = require("./api/Image");
exports.Image = Image_1.Image;
var ImageComments_1 = require("./api/ImageComments");
exports.ImageComments = ImageComments_1.ImageComments;
var ImageRepresentations_1 = require("./api/ImageRepresentations");
exports.ImageRepresentations = ImageRepresentations_1.ImageRepresentations;
var Link_1 = require("./api/Link");
exports.Link = Link_1.Link;
var SearchResults_1 = require("./api/SearchResults");
exports.SearchResults = SearchResults_1.SearchResults;
var Tag_1 = require("./api/Tag");
exports.Tag = Tag_1.Tag;
var TagDetails_1 = require("./api/TagDetails");
exports.TagDetails = TagDetails_1.TagDetails;
var User_1 = require("./api/User");
exports.User = User_1.User;
var DefaultFilters_1 = require("./util/DefaultFilters");
exports.DefaultFilters = DefaultFilters_1.DefaultFilters;
